File: cantor-go/pkg/structure/heap/binary.go
package heap

import "fmt"

type Element interface {
	Integer | UnsignedInteger | float32 | float64 | ~string
}

type Integer interface {
	int | int8 | int16 | int32 | int64
}

type UnsignedInteger interface {
	uint | uint8 | uint16 | uint32 | uint64
}

type BinaryHeap[T any] struct {
	elements   []T
	comparable func(a, b T) bool
	size       uint
}

// NewBinaryHeap creates a new binary heap (min heap by default)
func NewBinaryHeap[T Element](comparable func(a, b T) bool) *BinaryHeap[T] {
	if comparable == nil {
		comparable = func(a, b T) bool {
			return a < b
		}
	}
	return &BinaryHeap[T]{elements: make([]T, 0), comparable: comparable, size: 0}
}

func (heap *BinaryHeap[T]) Size() uint {
	return heap.size
}

func (heap *BinaryHeap[T]) IsEmpty() bool {
	return heap.size == 0
}

func (heap *BinaryHeap[T]) Peek() (*T, error) {
	if heap.IsEmpty() {
		return nil, fmt.Errorf("heap is empty")
	}
	return &heap.elements[0], nil
}

func (heap *BinaryHeap[T]) swap(i, j uint) {
	heap.elements[i], heap.elements[j] = heap.elements[j], heap.elements[i]
}

func (heap *BinaryHeap[T]) Insert(element T) {
	heap.size++
	heap.elements = append(heap.elements, element)
	index := heap.size - 1
	parentIndex := index / 2
	for index > 0 && heap.comparable(heap.elements[index], heap.elements[parentIndex]) {
		heap.swap(index, parentIndex)
		index = parentIndex
		parentIndex = index / 2
	}
}

func (heap *BinaryHeap[T]) Delete() *T {
	if heap.IsEmpty() {
		return nil
	}
	element := heap.elements[0]
	heap.swap(0, heap.size-1)
	heap.size--
	heap.elements = heap.elements[:heap.size]
	index := uint(0)
	for index < heap.size/2 {
		leftChildIndex := index*2 + 1
		rightChildIndex := index*2 + 2

		if rightChildIndex < heap.size && heap.comparable(heap.elements[rightChildIndex], heap.elements[leftChildIndex]) && heap.comparable(heap.elements[rightChildIndex], heap.elements[index]) {
			heap.swap(index, rightChildIndex)
			index = rightChildIndex
		} else if heap.comparable(heap.elements[leftChildIndex], heap.elements[index]) {
			heap.swap(index, leftChildIndex)
			index = leftChildIndex
		} else {
			break
		}
	}
	return &element
}

func (heap *BinaryHeap[T]) Sort() *[]T {
	var newSlice []T
	for {
		if heap.IsEmpty() {
			break
		}
		newSlice = append(newSlice, *heap.Delete())
	}
	return &newSlice
}

File: cantor-rb/lib/cantor/structure/heap/binary.rb
# frozen_string_literal: true

module Structure
  module Heap
    # Defines a Binary heap (default is min heap)
    class BinaryHeap
      attr_reader :size

      def initialize(comparable = nil)
        @elements = []
        @comparable = comparable || ->(a, b) { a < b }
        @size = 0
      end

      def empty?
        @size.zero?
      end

      def peek
        return nil if empty?

        @elements[0]
      end

      def swap(idx, idy)
        @elements[idx], @elements[idy] = @elements[idy], @elements[idx]
      end

      def insert(element)
        @size += 1
        @elements << element
        idx = @size - 1
        parent_idx = idx / 2

        while idx.positive? && @comparable.call(@elements[idx], @elements[parent_idx])
          swap(idx, parent_idx)
          idx = parent_idx
          parent_idx = idx / 2
        end
      end

      def delete
        return nil if empty?

        element = @elements[0]
        swap(0, @size - 1)
        @size -= 1
        @elements = @elements[0...@size]
        idx = 0

        while idx < @size / 2
          left = idx * 2 + 1
          right = idx * 2 + 2

          right_with_left = @comparable.call(@elements[right], @elements[left])
          right_with_idx = @comparable.call(@elements[right], @elements[idx])
          left_with_idx = @comparable.call(@elements[left], @elements[idx])
          if right < @size && right_with_left && right_with_idx
            swap(idx, right)
            idx = right
          elsif left_with_idx
            swap(idx, left)
            idx = left
          else
            break
          end
        end
        element
      end

      def sort
        array = []
        array << delete until @size.zero?
        array
      end
    end
  end
end
